File: crates/deepsec-cli/src/commands/metrics.rs
use crate::config_loader::Context;
use anyhow::Result;
use clap::Args as ClapArgs;
use colored::Colorize;
use deepsec_core::{RevalidationVerdict, runs::list_runs, store::load_all_file_records};
use std::collections::BTreeMap;

#[derive(Debug, ClapArgs)]
pub struct Args {
    #[arg(long)]
    pub project_id: String,
}

pub fn run(args: Args, ctx: &Context) -> Result<()> {
    let records = load_all_file_records(&ctx.data_root, &args.project_id)?;
    let runs = list_runs(&ctx.data_root, &args.project_id)?;

    let mut total_cost = 0.0;
    let mut total_in = 0u64;
    let mut total_out = 0u64;
    let mut total_duration = 0u64;
    for r in &runs {
        total_cost += r.stats.total_cost_usd.unwrap_or(0.0);
        total_in += r.stats.total_input_tokens.unwrap_or(0);
        total_out += r.stats.total_output_tokens.unwrap_or(0);
        total_duration += r.stats.total_duration_ms.unwrap_or(0);
    }

    let mut by_slug: BTreeMap<String, (usize, usize, usize)> = BTreeMap::new();
    let mut tp = 0;
    let mut fp = 0;
    let mut fx = 0;
    let mut un = 0;
    let mut total = 0;

    for rec in &records {
        for f in &rec.findings {
            total += 1;
            let entry = by_slug.entry(f.vuln_slug.clone()).or_default();
            entry.0 += 1;
            if let Some(rv) = &f.revalidation {
                match rv.verdict {
                    RevalidationVerdict::TruePositive => {
                        tp += 1;
                        entry.1 += 1;
                    }
                    RevalidationVerdict::FalsePositive => {
                        fp += 1;
                        entry.2 += 1;
                    }
                    RevalidationVerdict::Fixed => fx += 1,
                    RevalidationVerdict::Uncertain => un += 1,
                    RevalidationVerdict::AcceptedRisk => tp += 1,
                }
            }
        }
    }

    println!("{}", "metrics".bold());
    println!("  runs:            {}", runs.len());
    println!("  total cost:      ${:.4}", total_cost);
    println!("  input tokens:    {total_in}");
    println!("  output tokens:   {total_out}");
    println!("  total runtime:   {:.1}s", total_duration as f64 / 1000.0);
    println!();
    println!("  findings:        {total}");
    println!("    TP:            {tp}");
    println!("    FP:            {fp}");
    println!("    fixed:         {fx}");
    println!("    uncertain:     {un}");
    println!();
    println!("  by slug:");
    let mut entries: Vec<_> = by_slug.iter().collect();
    entries.sort_by(|a, b| b.1.0.cmp(&a.1.0));
    for (slug, (n, tp, fp)) in entries.into_iter().take(30) {
        println!("    {slug:<42} total={n:<5} TP={tp:<4} FP={fp:<4}");
    }
    Ok(())
}

File: internal/cli/commands/metrics.go
package commands

import (
	"fmt"
	"sort"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/noeljackson/deepsec/internal/core"
	"github.com/spf13/cobra"
)

// NewMetricsCmd aggregates metrics across all runs.
func NewMetricsCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var projectID string
	cmd := &cobra.Command{
		Use:   "metrics",
		Short: "Aggregate metrics across all runs",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx, err := loader()
			if err != nil {
				return err
			}
			records, err := ctx.DataRoot.LoadAllFileRecords(projectID)
			if err != nil {
				return err
			}
			runs, err := ctx.DataRoot.ListRuns(projectID)
			if err != nil {
				return err
			}

			var totalCost float64
			var totalIn, totalOut, totalDur uint64
			for _, r := range runs {
				if r.Stats.TotalCostUSD != nil {
					totalCost += *r.Stats.TotalCostUSD
				}
				if r.Stats.TotalInputTokens != nil {
					totalIn += *r.Stats.TotalInputTokens
				}
				if r.Stats.TotalOutputTokens != nil {
					totalOut += *r.Stats.TotalOutputTokens
				}
				if r.Stats.TotalDurationMs != nil {
					totalDur += *r.Stats.TotalDurationMs
				}
			}

			type slugStats struct{ total, tp, fp int }
			bySlug := map[string]*slugStats{}
			var tp, fp, fixed, uncertain, total int
			for _, rec := range records {
				for _, f := range rec.Findings {
					total++
					s := bySlug[f.VulnSlug]
					if s == nil {
						s = &slugStats{}
						bySlug[f.VulnSlug] = s
					}
					s.total++
					if f.Revalidation != nil {
						switch f.Revalidation.Verdict {
						case core.VerdictTruePositive, core.VerdictAcceptedRisk:
							tp++
							s.tp++
						case core.VerdictFalsePositive:
							fp++
							s.fp++
						case core.VerdictFixed:
							fixed++
						case core.VerdictUncertain:
							uncertain++
						}
					}
				}
			}

			fmt.Println("metrics")
			fmt.Printf("  runs:            %d\n", len(runs))
			fmt.Printf("  total cost:      $%.4f\n", totalCost)
			fmt.Printf("  input tokens:    %d\n", totalIn)
			fmt.Printf("  output tokens:   %d\n", totalOut)
			fmt.Printf("  total runtime:   %.1fs\n", float64(totalDur)/1000.0)
			fmt.Println()
			fmt.Printf("  findings:        %d\n", total)
			fmt.Printf("    TP:            %d\n", tp)
			fmt.Printf("    FP:            %d\n", fp)
			fmt.Printf("    fixed:         %d\n", fixed)
			fmt.Printf("    uncertain:     %d\n", uncertain)
			fmt.Println()
			fmt.Println("  by slug:")
			slugs := make([]string, 0, len(bySlug))
			for s := range bySlug {
				slugs = append(slugs, s)
			}
			sort.Slice(slugs, func(i, j int) bool {
				return bySlug[slugs[i]].total > bySlug[slugs[j]].total
			})
			limit := 30
			if limit > len(slugs) {
				limit = len(slugs)
			}
			for _, slug := range slugs[:limit] {
				s := bySlug[slug]
				fmt.Printf("    %-42s total=%-5d TP=%-4d FP=%-4d\n", slug, s.total, s.tp, s.fp)
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&projectID, "project-id", "", "Project id (required)")
	_ = cmd.MarkFlagRequired("project-id")
	return cmd
}
